src/main/java/nextstep/security/filter/FilterChainProxy.java

```diff
@@ -0,0 +1,36 @@
+package nextstep.security.filter;
+
+import jakarta.servlet.*;
+import jakarta.servlet.http.HttpServletRequest;
+import org.springframework.web.filter.GenericFilterBean;
+
+import java.io.IOException;
+import java.util.Collections;
+import java.util.List;
+import java.util.Optional;
+
+public class FilterChainProxy extends GenericFilterBean {
+    private final List<SecurityFilterChain> securityFilterChains;
+
+    public FilterChainProxy(final List<SecurityFilterChain> securityFilterChains) {
+        this.securityFilterChains = securityFilterChains;
+    }
+
+    @Override
+    public void doFilter(final ServletRequest request, final ServletResponse response, final FilterChain chain) throws IOException, ServletException {
+        List<Filter> filters = getFilters((HttpServletRequest) request);
+        var virtualFilterChain = new VirtualFilterChain(filters, chain);
+        virtualFilterChain.doFilter(request, response);
+    }
+
+    private List<Filter> getFilters(HttpServletRequest request) {
+        Optional<SecurityFilterChain> filters = securityFilterChains.stream()
+                .filter(it -> it.matches(request))
+                .findFirst();
+        if(filters.isEmpty()) {
+            return Collections.emptyList();
+        }
+        return filters.get().getFilters();
```

Comment:
```suggestion
        return securityFilterChains.stream()
                .filter(it -> it.matches(request))
                .findFirst()
                .map(SecurityFilterChain::getFilters)
                .orElse(Collections.emptyList());
```
이렇게 표현되는 것이 좀 더 Optional을 안전하게 사용할 수 있을 것 같네요 :)

Reply:
이 부분 안 그래도 부자연스럽다고 생각했는데 보완 감사드립니다!
제가 자바를 안 쓴지 오래돼서 Optional이랑 nullable한 객체들이 특히 어렵네요..